File: frontend/src/utils/rollBagUnits.ts
/**
 * Inventory for many items is in "rolls" (item UOM = ea) while pack master data
 * describes bags per roll (e.g. 175 pcs per roll). POs stay in rolls; production
 * can allocate in individual bags by converting bags ↔ rolls here.
 */

export interface PackSizeLike {
  pack_size: number
  pack_size_unit: string
}

/** Bags per one inventory ea (roll), from ItemPackSize on the lot. */
export function getBagsPerRollFromPackSize(pack: PackSizeLike | null | undefined): number | null {
  if (!pack || pack.pack_size == null || pack.pack_size <= 0) return null
  let u = (pack.pack_size_unit || '').trim().toLowerCase()
  if (!u) u = 'pcs'
  // 175 pcs = 175 bags per roll; some masters use ea for "units per roll"
  if (u === 'pcs' || u === 'pc' || u === 'ea' || u === 'piece' || u === 'pieces') return pack.pack_size
  return null
}

export function getBagsPerInventoryEa(lot: {
  item: { unit_of_measure: string }
  pack_size_obj?: PackSizeLike | null
}): number | null {
  if (lot.item.unit_of_measure !== 'ea') return null
  return getBagsPerRollFromPackSize(lot.pack_size_obj ?? undefined)
}

/** Lot API may omit pack_size_obj; use item default_pack_size / pack_sizes (pcs/ea = bags per roll). */
export function resolveBagsPerRollForLot(lot: {
  pack_size_obj?: PackSizeLike | null
  item?: {
    unit_of_measure?: string
    pack_sizes?: PackSizeLike[]
    default_pack_size?: PackSizeLike | null
  }
}): number | null {
  const uom = (lot.item?.unit_of_measure || '').trim().toLowerCase()
  if (uom !== 'ea') return null
  const fromLot = getBagsPerRollFromPackSize(lot.pack_size_obj ?? undefined)
  if (fromLot != null) return fromLot
  const def = lot.item?.default_pack_size
  if (def) {
    const b = getBagsPerRollFromPackSize(def)
    if (b != null) return b
  }
  for (const ps of lot.item?.pack_sizes ?? []) {
    const b = getBagsPerRollFromPackSize(ps)
    if (b != null) return b
  }
  return null
}

/** API may send bags_per_roll as string; JSON sometimes stringifies numbers. */
export function coercePositiveBagsPerRoll(value: unknown): number | null {
  if (value == null || value === '') return null
  const n = typeof value === 'number' ? value : parseFloat(String(value).replace(/,/g, ''))
  if (!Number.isFinite(n) || n <= 0) return null
  return n
}

export function bagsToRolls(bags: number, bagsPerRoll: number): number {
  if (bagsPerRoll <= 0) return 0
  return Math.round((bags / bagsPerRoll) * 1e6) / 1e6
}

export function rollsToBags(rolls: number, bagsPerRoll: number): number {
  return Math.round(rolls * bagsPerRoll * 1000) / 1000
}

File: backend_django/erp_core/roll_bag_units.py
"""
Roll/bag math for inventory and costing (mirrors frontend src/utils/rollBagUnits.ts).

Inventory for many packaging SKUs is stored in rolls (item UOM = ea). ItemPackSize
with unit pcs or ea gives bags per roll; actual bags on hand = rolls * bags_per_roll.
"""


def bags_per_roll_from_pack_size(pack):
    """Return bags per one inventory roll, or None if pack does not define that."""
    if pack is None:
        return None
    try:
        v = float(pack.pack_size)
    except (TypeError, ValueError):
        return None
    if v <= 0:
        return None
    u = (getattr(pack, 'pack_size_unit', None) or '').strip().lower()
    # Some rows omit unit; for roll stock + pack master numeric, treat as pcs/bags per roll
    if not u:
        u = 'pcs'
    if u in ('pcs', 'pc', 'ea', 'piece', 'pieces'):
        return v
    return None


def rolls_to_bags(rolls, bags_per_roll):
    return round(float(rolls) * float(bags_per_roll) * 1000) / 1000


def bags_to_rolls(bags, bags_per_roll):
    if bags_per_roll <= 0:
        return 0.0
    return round((float(bags) / float(bags_per_roll)) * 1e6) / 1e6


def unit_cost_per_bag(price_per_roll, bags_per_roll):
    """When price is per roll and pack master gives bags/roll, cost per bag."""
    if bags_per_roll is None or bags_per_roll <= 0:
        return None
    try:
        pr = float(price_per_roll)
    except (TypeError, ValueError):
        return None
    return pr / float(bags_per_roll)
